server/src/main/java/org/elasticsearch/node/NodeConstructor.java

```diff
@@ -0,0 +1,1634 @@
+import org.elasticsearch.transport.Transport;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.upgrades.SystemIndexMigrationExecutor;
+import org.elasticsearch.usage.UsageService;
+import org.elasticsearch.watcher.ResourceWatcherService;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.common.util.CollectionUtils.concatLists;
+import static org.elasticsearch.core.Types.forciblyCast;
+
+/**
+ * Class uses to perform all the operations needed to construct a {@link Node} instance.
+ */
+class NodeConstructor {
+
+    /**
+     * Performs node construction operations
+     *
+     * @param initialEnvironment         the initial environment for this node, which will be added to by plugins
+     * @param serviceProvider            provides various service implementations that could be mocked
+     * @param forbidPrivateIndexSettings whether or not private index settings are forbidden when creating an index; this is used in the
+     *                                   test framework for tests that rely on being able to set private settings
+     */
+    static NodeConstructor construct(
+        Environment initialEnvironment,
+        NodeServiceProvider serviceProvider,
+        boolean forbidPrivateIndexSettings
+    ) {
+        List<Closeable> closeables = new ArrayList<>();
+        try {
+            var constructor = new NodeConstructor(closeables);
+            constructor.checkEnvironment(initialEnvironment);
+            constructor.createEnvironment(initialEnvironment, serviceProvider.pluginsServiceCtor(initialEnvironment));
+            constructor.createThreadPools();
+            constructor.createTracer();
+            constructor.createTaskManager();
+            constructor.loadAdditionalSettings();
+            constructor.createClient();
+            constructor.createScriptService(serviceProvider);
+            constructor.createAnalysisModule();
+            // this is as early as we can validate settings at this point. we already pass them to ScriptModule as well as ThreadPool
+            // so we might be late here already
+            constructor.createSettingsModule();
+            // creating NodeEnvironment breaks the ability to rollback to 7.x on an 8.0 upgrade (upgradeLegacyNodeFolders) so do this
+            // after settings validation.
+            constructor.createNodeEnvironment();
```

Comment:
I understand the logic this is replacing is already incredibly dense, and so moving these constructions out to methods seems sensible. Yet there are interdependencies here which now cannot be seen (eg certain services which must be created before others). While I do hope we can work towards more well defined groupings of constructions, similar to Elasticsearch init phases, I'm not sure moving these out to individual methods is actually making things more understandable here.

For now, could we move the contents of the Node ctor entirely here, with all of these services, modules, etc as locals as they are now? Then we can work on regrouping and reordering in more manageable chunks.

Reply:
This is only the first step - there's a whole load of subsequent refactorings to trim this down which I've already got lined up, straight after this goes in.

I've done this initial split into the lowest groups to add information on which objects are only needed for one service (only used in that service method) and which are needed by multiple (so assigned a member field). I haven't changed the order, so the dependencies are unchanged.

The one thing this does affect is that it's now possible to call the methods in a different order, as the fields are always available - previously the variables were not defined & available until the services were constructed. To maintain this here, I could add non-null precondition checks to the start of each method, which declares its object dependencies? That would then make it very clear which method depends on what, and fail-fast if they are called in the wrong order.

Reply:
> The one thing this does affect is that it's now possible to call the methods in a different order

Maybe you can enforce it by using a stepped builder pattern (or an adaptation): you group interdependencies and hide them behind separate interfaces, and then use a fluent-style API to "progress" through steps.
(That's also part of the reason why I suggested to break down NodeFactory/FactoryBuilder - these steps seem to me to be a way of configuring the factory that will then be used to create a Node, following always the same steps).

E.g.: (totally inventing ordering here)

```
interface BuilderStep1 {
   // All this are order-independent, but do not make "progress" to the next phase possible
   BuilderStep1 createEnvironment(...);
   BuilderStep1 createThreadPools();
   ...
   // This method(s) make it move to the next phase
   BuilderStep2 loadAdditionalSettings();
} 

interface BuilderStep2 {
    BuilderStep2 createClient();
    BuilderStep2 createScriptService(serviceProvider);
    BuilderStep3 createAnalysisModule(); // etc.
}

// Last step Builder
... NodeConstructor build();

interface NodeConstructor {
    Client client();
    // etc.
}
```
Inside the NodeConstructor/NodeFactory
```
var constructor = builder
   .createEnvironment(...)
   .createThreadPools()
   .loadAdditionalSettings() // forces next step
   .createClient()
   ....
   .build();

return new Node(constructor.client(), ...);
```

Reply:
That's a good end goal, but we're not there yet. Hence this first step of separating out all the pieces we've got, so we can later put them back together in a more sensible way.

Reply:
My hesitation is around being able to "see" the patterns/groups which can be extracted. By behind hidden behind opaque methods, it's more difficult to tell what is being done, what can be elided, what could be local, etc. I would rather keep the method as-is, move it wholesale, then start trying to chip away pieces into separate groups/phases.

Reply:
I've reverted back to a single method copied from `Node`